src/Confluent.Kafka/Impl/SafeKafkaHandle.cs

```diff
@@ -492,54 +492,107 @@ internal ErrorCode PollSetConsumer()
 
         internal void CommitTransaction(int millisecondsTimeout)
         {
-            var errorStringBuilder = new StringBuilder(Librdkafka.MaxErrorStringLength);
-            var errorCode = Librdkafka.commit_transaction(this.handle, millisecondsTimeout, errorStringBuilder, (UIntPtr)errorStringBuilder.Capacity);
-            if (errorCode != ErrorCode.NoError)
+            var error = new Error(Librdkafka.commit_transaction(this.handle, (IntPtr)millisecondsTimeout));
+            if (error.Code != ErrorCode.NoError)
             {
-                throw new KafkaException(CreatePossiblyFatalError(errorCode, errorStringBuilder.ToString()));
+                if (error.TxnRequiresAbort)
+                {
+                    throw new KafkaTxnRequiresAbortException(error);
+                }
+                throw new KafkaException(error);
             }
         }
 
         internal void AbortTransaction(int millisecondsTimeout)
         {
-            var errorStringBuilder = new StringBuilder(Librdkafka.MaxErrorStringLength);
-            var errorCode = Librdkafka.abort_transaction(this.handle, millisecondsTimeout, errorStringBuilder, (UIntPtr)errorStringBuilder.Capacity);
-            if (errorCode != ErrorCode.NoError)
+            var error = new Error(Librdkafka.abort_transaction(this.handle, (IntPtr)millisecondsTimeout));
+            if (error.Code != ErrorCode.NoError)
             {
-                throw new KafkaException(CreatePossiblyFatalError(errorCode, errorStringBuilder.ToString()));
+                throw new KafkaException(error);
             }
         }
 
-        internal void SendOffsetsToTransaction(IEnumerable<TopicPartitionOffset> offsets, string group, int millisecondsTimeout)
+        internal void SendOffsetsToTransaction(IEnumerable<TopicPartitionOffset> offsets, IConsumerGroupMetadata groupMetadata, int millisecondsTimeout)
         {
             IntPtr offsetsPtr = GetCTopicPartitionList(offsets);
 
-            var errorStringBuilder = new StringBuilder(Librdkafka.MaxErrorStringLength);
-            var errorCode = Librdkafka.send_offsets_to_transaction(this.handle, offsetsPtr, group, millisecondsTimeout, errorStringBuilder, (UIntPtr)errorStringBuilder.Capacity);
-            if (errorCode != ErrorCode.NoError)
+            if (!(groupMetadata is ConsumerGroupMetadata))
+            {
+                throw new ArgumentException("groupMetadata object must be a value acquired via Consumer.ConsumerGroupMetadata.");
+            }
+            var serializedMetadata = ((ConsumerGroupMetadata)groupMetadata).serializedMetadata;
+            var cgmdPtr = this.DeserializeConsumerGroupMetadata(serializedMetadata);
+            try
+            {
+                var error = new Error(Librdkafka.send_offsets_to_transaction(this.handle, offsetsPtr, cgmdPtr, (IntPtr)millisecondsTimeout));
+                if (error.Code != ErrorCode.NoError)
+                {
+                    if (error.TxnRequiresAbort)
```

Comment:
Did we decide not to check retriable?

Reply:
I wasn't going to, but you made me realize there's no reason not to. Done in exactly the same way as txn-requires-abort.